OCPBUGS-3186: IBMCloud: Confirm Zones and BYON Subnets

cjschaef opened the pull request:
Confirm the specified Availability Zones, if any, properly map to the specified BYON Subnets, if any, on IBM Cloud. This confirms the specified values match and can be properly used during Control Plane and/or Compute deployments.

Related: https://issues.redhat.com/browse/OCPBUGS-3186

cjschaef commented:
/retitle OCPBUGS-3186: IBMCloud: Confirm Zones and BYON Subnets

r4f4 commented:
/jira refresh

MayXuQQ commented:
have not specify the zones, got exception
yq '.controlPlane.platform.ibmcloud.zones,.platform.ibmcloud.controlPlaneSubnets' ./install-config.yaml
null
[
  "maxu-3p2-control-plane-eu-gb-1",
  "maxu-3p2-control-plane-eu-gb-2"
]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x2b77a5b]

goroutine 1 [running]:
github.com/openshift/installer/pkg/asset/installconfig/ibmcloud.validateExistingSubnets({0x1a6f0a18, 0xc000e6eb10}, 0xc000da6a00, 0xc000e6ee70, {0xc00105f020, 0x29})
	/go/src/github.com/openshift/installer/pkg/asset/installconfig/ibmcloud/validation.go:288 +0x13b
github.com/openshift/installer/pkg/asset/installconfig/ibmcloud.validateExistingVPC({0x1a6f0a18, 0xc000e6eb10}, 0xc000da6a00, 0xc000e6ee70)

rvanderp3 reviewed (commented):
/lgtm

rvanderp3 commented:
/lgtm cancel

I just noticed the comment from @MayXuQQ

cjschaef commented:
@MayXuQQ  Correct me if I'm wrong,

I believe the [error above](https://github.com/openshift/installer/pull/6553#issuecomment-1303072938) is when the machinepoolplatform is not set or empty (`controlPlane.platform`)
```
# yq '{"controlPlane": .controlPlane, "platform": .platform}' byon-az-test-1/install-config.yaml 
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
  replicas: 3
platform:
  ibmcloud:
    region: ca-tor
    resourceGroupName: ca-tor-existing-network-1
    vpcName: ca-tor-existing-network-1-vpc
    controlPlaneSubnets:
      - ca-tor-existing-network-1-cp-ca-tor-2
      - ca-tor-existing-network-1-cp-ca-tor-3
    computeSubnets:
      - ca-tor-existing-network-1-compute-ca-tor-1
      - ca-tor-existing-network-1-compute-ca-tor-2
      - ca-tor-existing-network-1-compute-ca-tor-3

# go/src/github.com/openshift/installer/bin/openshift-install create manifests --dir byon-az-test-1
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x2b746db]

goroutine 1 [running]:
github.com/openshift/installer/pkg/asset/installconfig/ibmcloud.validateExistingSubnets({0x1aae6938, 0xc000c92ab0}, 0xc000b48a00, 0xc000c92e10, {0xc00084de60, 0x29})
	/root/go/src/github.com/openshift/installer/pkg/asset/installconfig/ibmcloud/validation.go:288 +0x13b
...
```
same with `controlPlane.platform.ibmcloud` (or `compute` machinepools too)
```
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    ibmcloud:
  replicas: 3
platform:
  ibmcloud:
    region: ca-tor
    resourceGroupName: ca-tor-existing-network-1
    vpcName: ca-tor-existing-network-1-vpc
    controlPlaneSubnets:
      - ca-tor-existing-network-1-cp-ca-tor-2
      - ca-tor-existing-network-1-cp-ca-tor-3
    computeSubnets:
      - ca-tor-existing-network-1-compute-ca-tor-1
      - ca-tor-existing-network-1-compute-ca-tor-2
      - ca-tor-existing-network-1-compute-ca-tor-3
```

I've updated the PR to now handle these cases, and added unit tests as well.
```
# yq '{"controlPlane": .controlPlane, "compute": .compute[0], "platform": .platform}' byon-az-test-1/install-config.yaml 
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    ibmcloud:
  replicas: 3
compute:
  architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform:
  replicas: 3
platform:
  ibmcloud:
    region: ca-tor
    resourceGroupName: ca-tor-existing-network-1
    vpcName: ca-tor-existing-network-1-vpc
    controlPlaneSubnets:
      - ca-tor-existing-network-1-cp-ca-tor-2
      - ca-tor-existing-network-1-cp-ca-tor-3
    computeSubnets:
      - ca-tor-existing-network-1-compute-ca-tor-1
      - ca-tor-existing-network-1-compute-ca-tor-2

# go/src/github.com/openshift/installer/bin/openshift-install create manifests --dir byon-az-test-1
ERROR failed to fetch Master Machines: failed to load asset "Install Config": failed to create install config: [platform.ibmcloud.controlPlaneSubnets: Invalid value: []string{"ca-tor-existing-network-1-cp-ca-tor-2", "ca-tor-existing-network-1-cp-ca-tor-3"}: number of zones (2) covered by controlPlaneSubnets does not match number of provided or default zones (3) for control plane in ca-tor, platform.ibmcloud.computeSubnets: Invalid value: []string{"ca-tor-existing-network-1-compute-ca-tor-1", "ca-tor-existing-network-1-compute-ca-tor-2"}: number of zones (2) covered by computeSubnets does not match number of provided or default zones (3) for compute[0] in ca-tor]
```

cjschaef commented:
`e2e-ibmcloud-ovn` affected by https://issues.redhat.com/browse/OCPBUGS-1327
```
openshift-machine-api   ci-op-zwy4qjz1-52953-6vdn9-worker-2-nhktd   Provisioned   bx2-4x16   us-east   us-east-2   35m                                               ibm://b9db97fa92ce47ae8dbf40ec2318b9f4///ci-op-zwy4qjz1-52953-6vdn9/0767_a87c091b-b0d5-4eaf-ad94-2808269b983a   running
```
https://github.com/openshift/machine-api-provider-ibmcloud/pull/10 has merged, so I suspect we will soon see builds with this mitigation to help improve CI results.

MayXuQQ reviewed (approved):
pre-merge test based on  #6553 cd078e85 registry.build05.ci.openshift.org/ci-ln-34cjr4k/release:latest

cjschaef commented:
`e2e-ibmcloud-ovn` looks like an unexpected error during build, unrelated to these changes.
```
error: build error: error building at STEP "RUN yum update -y &&     yum install --setopt=tsflags=nodocs -y     python3-openstackclient ansible-2.9.14-1.el8ae python3-openstacksdk python3-netaddr unzip jq &&     yum clean all && rm -rf /var/cache/yum/*": error while running runtime: exit status 1
```
Trying again
/retest

cjschaef commented:
`e2e-ibmcloud-ovn` failed with another build error
```
 ERRO[2022-11-08T21:18:23Z] Some steps failed:                           
ERRO[2022-11-08T21:18:23Z] 
  * could not run steps: step openstack-installer failed: the build openstack-installer failed after 18m44s with reason DockerBuildFailed: Dockerfile build strategy has failed. 
INFO[2022-11-08T21:18:23Z] Reporting job state 'failed' with reason 'executing_graph:step_failed:building_project_image' 
```
Trying again
/retest

rvanderp3 commented:
/lgtm

r4f4 reviewed (commented):
/approve

cjschaef commented:
/cherrypick release-4.12
